src/workerd/api/tcp.h

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#pragma once
+
+#include <workerd/jsg/jsg.h>
+#include "http.h"
+
+
+namespace workerd::api {
+
+class PipelinedAsyncIoStream final : public kj::AsyncIoStream, public kj::Refcounted {
+  // A stream that is backed by a promise for an AsyncIoStream. All operations on this stream
+  // are deferred until the `inner` promise completes.
+public:
+  explicit PipelinedAsyncIoStream(kj::Promise<kj::Own<kj::AsyncIoStream>> inner);
+  void shutdownWrite() override;
+  void abortRead() override;
+  void getsockopt(int level, int option, void* value, uint* length) override;
+  void setsockopt(int level, int option, const void* value, uint length) override;
+
+  void getsockname(struct sockaddr* addr, uint* length) override;
+  void getpeername(struct sockaddr* addr, uint* length) override;
+
+  // AsyncInputStream methods:
+  kj::Promise<size_t> read(void* buffer, size_t minBytes, size_t maxBytes) override;
+  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override;
+
+  // AsyncOutputStream methods:
+  kj::Promise<void> write(const void* buffer, size_t size) override;
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override;
+
+  kj::Promise<void> whenWriteDisconnected() override;
+
+private:
+  // Either calls `then` on the `inner` promise with the `f` callback, or calls `f` on an already
+  // stored stream (if the `inner` promise completed).
+  template <typename T>
+  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::Own<kj::AsyncIoStream> *)> f) {
+    KJ_IF_MAYBE(e, error) {
+      kj::throwRecoverableException(kj::mv(*e));
+    }
+
+    KJ_IF_MAYBE(s, completedInner) {
+      return f(s);
+    } else {
+      return inner.then([this, f](kj::Own<kj::AsyncIoStream> stream) -> kj::Promise<T> {
+        completedInner = kj::mv(stream);
+        return f(&KJ_ASSERT_NONNULL(completedInner));
+      });
+    }
+  }
+
+  // Stored io stream once the `inner` promise completes.
+  kj::Maybe<kj::Own<kj::AsyncIoStream>> completedInner;
+  // Stored error if any of the operations throw.
+  kj::Maybe<kj::Exception> error;
+  // A promise holding a stream that will be available at some point in the future. Typically once
+  // a connection is established.
+  kj::Promise<kj::Own<kj::AsyncIoStream>> inner;
+};
+
+class SocketOptions: public jsg::Object {
+  // TODO(later): TCP socket options will live here.
+  SocketOptions() {};
+public:
+  JSG_RESOURCE_TYPE(SocketOptions, CompatibilityFlags::Reader flags) { }
+};
+
+class Socket: public jsg::Object {
+public:
+  Socket(kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise);
+
+  kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise;
+
+  kj::Maybe<jsg::Ref<ReadableStream>> readable;
```

Comment:
Why make the `readable` and `writable` public like this? I'd prefer they be private.

Also, just to prepare for future, good idea to implement a `visitForGc` method that visits these.

Reply:
That's just an oversight on my part, will make these private.

Reply:
Done, and added `visitForGc` as well.